server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java

```diff
@@ -234,11 +236,11 @@ private boolean cancelTaskIfGranularityChanged(
     Granularity configuredSegmentGranularity = dataSourceCompactionConfig.getGranularitySpec()
                                                                          .getSegmentGranularity();
     Granularity taskSegmentGranularity = compactionTaskQuery.getGranularitySpec().getSegmentGranularity();
-    if (configuredSegmentGranularity.equals(taskSegmentGranularity)) {
+    if (configuredSegmentGranularity == null || configuredSegmentGranularity.equals(taskSegmentGranularity)) {
```

Comment:
This would surface when a previous compaction task is already in flight. Perhaps since I was running compaction duty frequently at 30s in my local testing, this came to the fore.